optuna_dashboard/ts/components/GraphTimeline.tsx

```diff
@@ -106,14 +101,38 @@ const plotTimeline = (trials: Trial[], mode: string) => {
       y: bars.map((b) => b.number),
       // @ts-ignore: To suppress ts(2322)
       base: starts.map((s) => s.toISOString()),
-      name: s,
+      name: name,
       text: bars.map((b) => makeHovertext(b)),
-      hovertemplate: "%{text}<extra>" + s + "</extra>",
+      hovertemplate: "%{text}<extra>" + name + "</extra>",
       orientation: "h",
-      marker: { color: cm[s] },
+      marker: { color: color },
       textposition: "none", // Avoid drawing hovertext in a bar.
     }
-    traces.push(trace)
+    return trace
+  }
+
+  const traces: Partial<plotly.PlotData>[] = []
+  for (const s of Object.keys(cm) as TrialState[]) {
+    const bars = trials.filter((t) => t.state === s)
+    if (bars.length === 0) {
+      continue
+    }
+    if (s == "Complete") {
+      const feasibleTrials = bars.filter((t) =>
+        t.constraints.every((c) => c <= 0)
+      )
+      const infeasibleTrials = bars.filter((t) =>
+        t.constraints.some((c) => c > 0)
+      )
+      if (feasibleTrials.length > 0) {
+        traces.push(makeTrace(feasibleTrials, "Complete", cm[s]))
+      }
+      if (infeasibleTrials.length > 0) {
+        traces.push(makeTrace(infeasibleTrials, "Infeasible", "#cccccc"))
+      }
+    } else {
+      traces.push(makeTrace(bars, s, cm[s]))
+    }
```

Comment:
What about early return?

```suggestion
    if (s !== "Complete") {
      traces.push(makeTrace(bars, s, cm[s]))
      continue
    }
    const feasibleTrials = bars.filter((t) =>
      t.constraints.every((c) => c <= 0)
    )
    const infeasibleTrials = bars.filter((t) =>
      t.constraints.some((c) => c > 0)
    )
    if (feasibleTrials.length > 0) {
      traces.push(makeTrace(feasibleTrials, "Complete", cm[s]))
    }
    if (infeasibleTrials.length > 0) {
      traces.push(makeTrace(infeasibleTrials, "Infeasible", "#cccccc"))
    }
```

Reply:
I don't think early return is appropriate here since the codes inside the brackets are at the same level. For example, if we add another if statement for "Waiting", it should be like this.

```typescript
if (s === "Complete") {
  ...
} else if (s === "Waiting") {
  ...
} else {
  ...
}
```

not like this.

```typescript
if (s !== "Complete" && s !== "Waiting") {
  ...
  continue
}
if (s === "Complete") {
  ...
  continue
}
...
```